frontend/src/frontend_benchmark/user_callgraph.py

```diff
@@ -0,0 +1,200 @@
+        return self.functions[function_name]
+
+    def format_vars_declaration(self) -> str:
+        return self.format_code_block_body(self.global_vars_decl)
+
+    def format_headers(self) -> str:
+        headers = []
+        for function in self.functions.values():
+            headers.append(f'{function.get_signature_header()}();')
+        out = '\n'.join(headers)
+        out += '\n'
+        return out
+
+    def direct_call_targets_for_function(
+            self, function_name) -> Collection[Optional[int]]:
+        function = self.get_function(function_name)
+        return function.get_branch_targets(Branch.filter(
+            BranchType.DIRECT_CALL))
+
+    def format_function(self, function_name: int) -> str:
+        function = self.get_function(function_name)
+        result = function.get_signature_header() + '() {\n'
+        code_block_texts = [
+            self.format_code_block_with_label(code_block)
+            for code_block in function.code_blocks
+        ]
+        result = result + ''.join(code_block_texts)
+        result += '}\n'
+        return result
+
+    def format_code_block_with_label(self, codeblock: CodeBlock) -> str:
+        return (f'{self.format_code_block_label(codeblock)}:;\n'
+                f'{self.format_code_block(codeblock)}')
+
+    def format_code_block_label(self, codeblock: CodeBlock) -> str:
+        return f'label{codeblock.name}'
+
+    def format_code_block(self, codeblock: CodeBlock) -> str:
+        cbb_text = self.format_code_block_body(codeblock)
+        branch_text = self.format_branch(codeblock.terminator_branch)
+        return f'{cbb_text}{branch_text}'
+
+    def format_branch(self, branch: Branch) -> str:
+        branch_options = {
+            BranchType.INDIRECT_CALL:
+                self._format_branch_indirect_call,
+            BranchType.DIRECT_CALL:
+                self._format_branch_direct_call,
+            BranchType.FALLTHROUGH:
+                self._format_branch_fallthrough,
+            BranchType.UNKNOWN:
+                self._format_branch_fallthrough,
+            BranchType.CONDITIONAL_DIRECT:
+                self._format_branch_conditional_direct,
+            BranchType.CONDITIONAL_INDIRECT:
+                self._format_branch_conditional_indirect,
+            BranchType.INDIRECT:
+                self._format_branch_indirect,
+            BranchType.DIRECT:
+                self._format_branch_direct,
+            BranchType.RETURN:
+                self._format_branch_return
+        }
+        if branch.branch_type in branch_options:
+            return branch_options[branch.branch_type](branch)
+        raise ValueError(f'Unknown branch type: {branch.branch_type}')
+
+    def _format_branch_indirect_call(self, branch: Branch) -> str:
+        target = branch.next_valid_target()
+        string = self.function_call_signature_for(target)
+        return f'void (*frontend_f)(void) = {string};\nfrontend_f();\n'
```

Comment:
I think you need an & before the signature?

Also, I still owe you a benchmark with indirect branches. Dropped the ball on that one.

Reply:
I get the same result with and without the & before the signature. I will leave it as is and verify it works with the indirect branch test you provide.